app/models/service_template.rb

```diff
@@ -340,8 +353,77 @@ def provision_action
     resource_actions.find_by(:action => "Provision")
   end
 
+  def update_resource_actions(ae_endpoints)
+    resource_action_options.each do |action|
+      resource_action = resource_actions.find_by(:action => action[:name])
+      # If the action exists in updated parameters
+      if ae_endpoints[action[:param_key]]
+        # And the resource action exists on the template already, update it
+        if resource_action
+          resource_actions.find_by(:action => action[:name]).update_attributes!(ae_endpoints[action[:param_key]])
+        # If the resource action does not exist, create it
+        else
+          build_resource_action(ae_endpoints[action[:param_key]], action)
+        end
+      elsif resource_action
+        # If the endpoint does not exist in updated parameters, but exists on the template, delete it
+        resource_action.destroy
+      end
+    end
+  end
+
   def create_resource_actions(ae_endpoints)
     ae_endpoints ||= {}
+    resource_action_options.each do |action|
+      ae_endpoint = ae_endpoints[action[:param_key]]
+      next unless ae_endpoint
+      build_resource_action(ae_endpoint, action)
+    end
+    save!
+  end
+
+  def self.create_from_options(options)
+    create(options.except(:config_info).merge(:options => { :config_info => options[:config_info] }))
+  end
+  private_class_method :create_from_options
+
+  private
+
+  def update_service_resources(config_info, auth_user = nil)
+    config_info = config_info.except(:provision, :retirement, :reconfigure)
+    workflow_class = MiqProvisionWorkflow.class_for_source(config_info[:src_vm_id])
+    if workflow_class
+      service_resources.find_by(:resource_type => 'MiqRequest').destroy
+      new_request = workflow_class.new(config_info, auth_user).make_request(nil, config_info)
+
+      add_resource!(new_request)
+    end
+  end
+
+  def build_resource_action(ae_endpoint, action)
+    fqname = if ae_endpoint.empty?
+               self.class.send(action[:method], *action[:args]) || ""
+             else
+               ae_endpoint[:fqname]
+             end
+
+    build_options = {:action        => action[:name],
+                     :fqname        => fqname,
+                     :ae_attributes => {:service_action => action[:name]}}
+    build_options.merge!(ae_endpoint.slice(:dialog,
+                                           :dialog_id,
+                                           :configuration_template,
+                                           :configuration_template_id,
+                                           :configuration_template_type))
+    resource_actions.build(build_options)
+  end
+
+  def validate_update_config_info(options)
+    raise _('service_type and prov_type cannot be changed') if options[:service_type] || options[:prov_type]
+    options[:config_info]
+  end
+
+  def resource_action_options
```

Comment:
We use the word options every where. Can we come up a better name for this method? Something like `resource_action_list`. Naming is hard :(